File: scripts/ai-workflow/core/claude_agent_client.py
"""Claude Agent SDK クライアント

Claude Code headless modeを使ってAIエージェントを実行
- Claude Pro Max契約を活用
- Read/Write/Edit/Bash/Grep/Globツールが使える
- 自律的にタスクを実行
"""
import os
import anyio
from pathlib import Path
from typing import Optional, List, Dict
from claude_agent_sdk import query, ClaudeAgentOptions


class ClaudeAgentClient:
    """Claude Agent SDK クライアント"""

    def __init__(self, working_dir: Optional[Path] = None, model: Optional[str] = None):
        """
        初期化

        Args:
            working_dir: 作業ディレクトリ（省略時はカレントディレクトリ）
            model: 使用するClaudeモデル（省略時はClaude Code Pro Maxのデフォルトモデルを使用）
                  - CLAUDE_CODE_OAUTH_TOKEN 使用時: デフォルトで claude-sonnet-4-5-20250929（最新Sonnet）
                  - ANTHROPIC_API_KEY 使用時: 明示的に指定が必要
        """
        self.working_dir = working_dir or Path.cwd()
        # CLAUDE_CODE_OAUTH_TOKEN を使う場合は強力なモデルを使用
        # model が指定されていない場合は、Claude Code Pro Max のデフォルトモデルを使用
        self.model = model

    async def execute_task(
        self,
        prompt: str,
        system_prompt: Optional[str] = None,
        max_turns: int = 50,
        verbose: bool = True
    ) -> List[str]:
        """
        タスクを実行

        Args:
            prompt: タスクプロンプト
            system_prompt: システムプロンプト
            max_turns: 最大ターン数
            verbose: 詳細ログ出力（リアルタイムメッセージ表示）

        Returns:
            List[str]: レスポンスメッセージのリスト

        Raises:
            ValueError: CLAUDE_CODE_OAUTH_TOKEN が設定されていない場合
        """
        # CLAUDE_CODE_OAUTH_TOKEN が設定されているか確認
        if not os.environ.get('CLAUDE_CODE_OAUTH_TOKEN'):
            raise ValueError(
                "CLAUDE_CODE_OAUTH_TOKEN が設定されていません。\n"
                "Claude Agent SDK は CLAUDE_CODE_OAUTH_TOKEN (Claude Code Pro Max) を使用します。\n"
                "設定方法: export CLAUDE_CODE_OAUTH_TOKEN='sk-ant-oat01-...'\n"
                "詳細は DOCKER_AUTH_SETUP.md を参照してください。"
            )

        # 環境変数でBashコマンド承認スキップを有効化（Docker環境内で安全）
        # Note: すでにDocker containerで隔離されているため、セキュリティリスクは限定的
        os.environ['CLAUDE_DANGEROUSLY_SKIP_PERMISSIONS'] = '1'

        print("[INFO] Using CLAUDE_CODE_OAUTH_TOKEN for Claude Agent SDK (Claude Code Pro Max)")

        # ClaudeAgentOptions の設定
        options_kwargs = {
            'system_prompt': system_prompt,
            'max_turns': max_turns,
            'cwd': str(self.working_dir),
            'permission_mode': 'bypassPermissions',  # すべてのツール使用を自動承認（Docker環境内で安全）
        }

        # model が指定されている場合のみ追加
        # デフォルト（model=None）の場合は Claude Code Pro Max のデフォルトモデル（Sonnet 4.5）を使用
        if self.model is not None:
            options_kwargs['model'] = self.model

        options = ClaudeAgentOptions(**options_kwargs)

        messages = []
        async for message in query(prompt=prompt, options=options):
            message_str = str(message)
            messages.append(message_str)

            # リアルタイムログ出力
            if verbose:
                # AssistantMessageの場合は思考内容を表示
                if 'AssistantMessage' in message_str:
                    # TextBlockを抽出して表示（全文表示）
                    if 'TextBlock(text=' in message_str:
                        start = message_str.find('TextBlock(text=') + 16
                        end = message_str.find('\')', start)
                        if end > start:
                            text = message_str[start:end]
                            print(f"[AGENT THINKING] {text}")
                    # ToolUseBlockを抽出して表示（詳細情報含む）
                    elif 'ToolUseBlock' in message_str:
                        if 'name=' in message_str:
                            # ツール名を抽出
                            name_start = message_str.find('name=') + 6
                            name_end = message_str.find('\'', name_start)
                            tool_name = message_str[name_start:name_end]

                            # 詳細ログを出力
                            print(f"[AGENT ACTION] Using tool: {tool_name}")

                            # ツールパラメータを抽出（input=以降、次のフィールドの手前まで）
                            if 'input=' in message_str:
                                params_start = message_str.find('input=') + 6
                                # ')' が次のフィールドの開始なので、そこまでを取得
                                # ToolUseBlock(id='...', name='...', input={...}) の構造
                                # input= の後から最後の ')' の手前までが input の内容

                                # より正確に抽出: input= から次の '), ' または末尾の ')' まで
                                # まず 'ToolUseBlock' の終わりを探す
                                toolblock_end = message_str.find(')]', params_start)
                                if toolblock_end == -1:
                                    toolblock_end = len(message_str)

                                params_str = message_str[params_start:toolblock_end].strip()
                                # 最後の ')' を除去
                                if params_str.endswith(')'):
                                    params_str = params_str[:-1]

                                # 長すぎる場合は省略（最初の500文字）
                                if len(params_str) > 500:
                                    params_str = params_str[:500] + "..."

                                print(f"[AGENT ACTION] Parameters: {params_str}")

        return messages

    async def execute_task_from_file(
        self,
        prompt_file: Path,
        template_vars: Optional[Dict[str, str]] = None,
        system_prompt: Optional[str] = None,
        max_turns: int = 50
    ) -> List[str]:
        """
        プロンプトファイルからタスクを実行

        Args:
            prompt_file: プロンプトファイルパス
            template_vars: テンプレート変数（{variable_name}を置換）
            system_prompt: システムプロンプト
            max_turns: 最大ターン数

        Returns:
            List[str]: レスポンスメッセージのリスト
        """
        # プロンプトファイルを読み込み
        prompt = prompt_file.read_text(encoding='utf-8')

        # テンプレート変数を置換
        if template_vars:
            for key, value in template_vars.items():
                prompt = prompt.replace(f'{{{key}}}', value)

        return await self.execute_task(
            prompt=prompt,
            system_prompt=system_prompt,
            max_turns=max_turns
        )

    def execute_task_sync(
        self,
        prompt: str,
        system_prompt: Optional[str] = None,
        max_turns: int = 50,
        verbose: bool = True
    ) -> List[str]:
        """
        タスクを同期実行（anyio.runを使用）

        Args:
            prompt: タスクプロンプト
            system_prompt: システムプロンプト
            max_turns: 最大ターン数
            verbose: 詳細ログ出力（リアルタイムメッセージ表示）

        Returns:
            List[str]: レスポンスメッセージのリスト
        """
        return anyio.run(
            self.execute_task,
            prompt,
            system_prompt,
            max_turns,
            verbose
        )

    def execute_task_from_file_sync(
        self,
        prompt_file: Path,
        template_vars: Optional[Dict[str, str]] = None,
        system_prompt: Optional[str] = None,
        max_turns: int = 50
    ) -> List[str]:
        """
        プロンプトファイルからタスクを同期実行

        Args:
            prompt_file: プロンプトファイルパス
            template_vars: テンプレート変数
            system_prompt: システムプロンプト
            max_turns: 最大ターン数

        Returns:
            List[str]: レスポンスメッセージのリスト
        """
        return anyio.run(
            self.execute_task_from_file,
            prompt_file,
            template_vars,
            system_prompt,
            max_turns
        )

File: scripts/ai-workflow-v2/src/core/claude-agent-client.ts
import fs from 'fs-extra';
import { query, type SDKMessage, type SDKAssistantMessage, type SDKResultMessage } from '@anthropic-ai/claude-agent-sdk';

interface ExecuteTaskOptions {
  prompt: string;
  systemPrompt?: string | null;
  maxTurns?: number;
  workingDirectory?: string;
  verbose?: boolean;
}

const DEFAULT_MAX_TURNS = 50;
const MAX_LOG_PARAM_LENGTH = 500;

export class ClaudeAgentClient {
  private readonly workingDir: string;
  private readonly model?: string;

  constructor(options: { workingDir?: string; model?: string; credentialsPath?: string } = {}) {
    this.workingDir = options.workingDir ?? process.cwd();
    this.model = options.model;

    this.ensureAuthToken(options.credentialsPath);

    // 環境変数の設定を確認
    const skipPermissions = process.env.CLAUDE_DANGEROUSLY_SKIP_PERMISSIONS === '1';
    if (skipPermissions) {
      console.info('[INFO] CLAUDE_DANGEROUSLY_SKIP_PERMISSIONS=1 detected. Using permissionMode="bypassPermissions".');
    } else {
      console.info('[INFO] Using permissionMode="acceptEdits" (default).');
    }
  }

  public getWorkingDirectory(): string {
    return this.workingDir;
  }

  public async executeTask(options: ExecuteTaskOptions): Promise<string[]> {
    const { prompt, systemPrompt = null, maxTurns = DEFAULT_MAX_TURNS, verbose = true } = options;
    const cwd = options.workingDirectory ?? this.workingDir;

    // 環境変数でBashコマンド承認スキップを確認（Docker環境内で安全）
    // CLAUDE_DANGEROUSLY_SKIP_PERMISSIONS=1 の場合、すべての操作を自動承認
    const skipPermissions = process.env.CLAUDE_DANGEROUSLY_SKIP_PERMISSIONS === '1';
    const permissionMode = skipPermissions ? 'bypassPermissions' : 'acceptEdits';

    const stream = query({
      prompt,
      options: {
        cwd,
        permissionMode,
        maxTurns,
        model: this.model,
        systemPrompt: systemPrompt ?? undefined,
      },
    });

    const messages: string[] = [];

    for await (const message of stream) {
      messages.push(JSON.stringify(message));
      if (verbose) {
        this.logMessage(message);
      }
    }

    return messages;
  }

  public async executeTaskFromFile(
    promptFile: string,
    templateVars?: Record<string, string>,
    systemPrompt?: string,
    maxTurns?: number,
    verbose?: boolean,
  ): Promise<string[]> {
    const template = fs.readFileSync(promptFile, 'utf-8');
    const prompt = this.fillTemplate(template, templateVars ?? {});
    return this.executeTask({ prompt, systemPrompt, maxTurns, verbose });
  }

  private fillTemplate(template: string, variables: Record<string, string>): string {
    let content = template;
    for (const [key, value] of Object.entries(variables)) {
      content = content.replace(new RegExp(`{${key}}`, 'g'), value);
    }
    return content;
  }

  private logMessage(message: SDKMessage): void {
    switch (message.type) {
      case 'assistant':
        this.logAssistantMessage(message);
        break;
      case 'result':
        this.logResultMessage(message);
        break;
      case 'system': {
        const subtype = 'subtype' in message && message.subtype ? message.subtype : 'system';
        console.log(`[AGENT SYSTEM] ${subtype}`);
        break;
      }
      case 'stream_event':
        this.logStreamEvent(message);
        break;
      default:
        break;
    }
  }

  private logAssistantMessage(message: SDKAssistantMessage): void {
    const contents = message.message.content ?? [];
    for (const block of contents) {
      if (block.type === 'text') {
        const text = block.text?.trim();
        if (text) {
          console.log(`[AGENT THINKING] ${text}`);
        }
      } else if (block.type === 'tool_use') {
        console.log(`[AGENT ACTION] Using tool: ${block.name}`);
        if (block.input && Object.keys(block.input).length > 0) {
          const raw = JSON.stringify(block.input);
          const truncated = raw.length > MAX_LOG_PARAM_LENGTH ? `${raw.slice(0, MAX_LOG_PARAM_LENGTH)}…` : raw;
          console.log(`[AGENT ACTION] Parameters: ${truncated}`);
        }
      }
    }
  }

  private logResultMessage(message: SDKResultMessage): void {
    console.log(
      `[AGENT RESULT] status=${message.subtype ?? 'success'}, turns=${message.num_turns}, duration_ms=${message.duration_ms}`,
    );
    const resultText = (message as Partial<{ result: string }>).result;
    if (typeof resultText === 'string' && resultText.trim().length > 0) {
      console.log(`[AGENT RESULT] ${resultText}`);
    }
  }

  private logStreamEvent(message: SDKMessage): void {
    if (message.type !== 'stream_event') {
      return;
    }

    const event = message.event;
    if (!event || event.type !== 'message_delta') {
      return;
    }

    const delta = event.delta;
    if (!delta || delta.type !== 'message_delta') {
      return;
    }

    for (const block of delta.delta?.content ?? []) {
      if (block.type === 'input_json_delta' && block.partial_input_json) {
        const raw = block.partial_input_json;
        const truncated = raw.length > MAX_LOG_PARAM_LENGTH ? `${raw.slice(0, MAX_LOG_PARAM_LENGTH)}…` : raw;
        console.log(`[AGENT ACTION] Partial parameters: ${truncated}`);
      } else if (block.type === 'text_delta' && block.text_delta?.trim()) {
        console.log(`[AGENT THINKING] ${block.text_delta.trim()}`);
      }
    }
  }

  private ensureAuthToken(credentialsPath?: string): void {
    const resolvedPath = credentialsPath ?? process.env.CLAUDE_CODE_CREDENTIALS_PATH ?? null;

    if (resolvedPath) {
      const token = this.readTokenFromCredentials(resolvedPath);
      console.info(`[INFO] Loaded Claude Code credentials from ${resolvedPath} (token length=${token.length})`);
      process.env.CLAUDE_CODE_OAUTH_TOKEN = token;
      return;
    }

    const token = process.env.CLAUDE_CODE_OAUTH_TOKEN;
    if (!token || !token.trim()) {
      throw new Error(
        [
          'Claude Code credentials are not configured.',
          'Provide a valid credentials file via CLAUDE_CODE_CREDENTIALS_PATH or set CLAUDE_CODE_OAUTH_TOKEN.',
        ].join('\n'),
      );
    }
  }

  private readTokenFromCredentials(credentialsPath: string): string {
    if (!fs.existsSync(credentialsPath)) {
      throw new Error(`Claude Code credentials file not found: ${credentialsPath}`);
    }

    const raw = fs.readFileSync(credentialsPath, 'utf-8').trim();
    if (!raw) {
      throw new Error(`Claude Code credentials file is empty: ${credentialsPath}`);
    }

    let token: string | null = null;
    try {
      const parsed = JSON.parse(raw);
      if (typeof parsed === 'string') {
        token = parsed.trim();
      } else {
        token = this.extractToken(parsed);
      }
    } catch {
      // Not JSON – treat as raw token string.
    }

    if (!token) {
      const trimmed = raw.trim();
      if (trimmed) {
        token = trimmed;
      }
    }

    if (!token) {
      throw new Error(`Unable to extract Claude Code token from credentials file: ${credentialsPath}`);
    }

    return token;
  }

  private extractToken(value: unknown): string | null {
    if (typeof value === 'string') {
      const trimmed = value.trim();
      if (trimmed && trimmed.length > 20 && !trimmed.includes(' ')) {
        return trimmed;
      }
      return null;
    }

    if (Array.isArray(value)) {
      for (const item of value) {
        const token = this.extractToken(item);
        if (token) {
          return token;
        }
      }
      return null;
    }

    if (value && typeof value === 'object') {
      const obj = value as Record<string, unknown>;
      const directKeys = ['token', 'access_token', 'accesstoken', 'oauth_token'];
      for (const key of Object.keys(obj)) {
        const candidate = obj[key];
        if (typeof candidate === 'string') {
          const lower = key.toLowerCase();
          if (directKeys.includes(lower)) {
            const trimmed = candidate.trim();
            if (trimmed) {
              return trimmed;
            }
          }
        }
      }

      for (const nested of Object.values(obj)) {
        const token = this.extractToken(nested);
        if (token) {
          return token;
        }
      }
    }

    return null;
  }

}
